src/filesystem/drive/inner.rs

```diff
@@ -48,11 +62,75 @@ impl InnerDrive {
             ))
     }
 
+    fn mark_ancestors_dirty(&mut self, node_id: NodeId) -> Result<(), OperationError> {
+        // Changes have happened in `node_id` walk up its parents to root marking nodes as dirty
+        let mut node_id = node_id;
+        self.dirty_nodes.push(node_id);
+        while let Some(parent_perm_id) = self.by_id(node_id)?.parent_id() {
+            // Setting a node's parent ID to its own ID is done in some operations such as moving a node
+            // to temporarily orphan it. Obviously this would cause an infinite loop and needs to be caught
+            if parent_perm_id == self.by_id(node_id)?.permanent_id() {
+                break;
+            }
+            node_id = self.lookup_internal_id(&parent_perm_id)?;
+            self.dirty_nodes.push(node_id);
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn clean_drive(&mut self) {
+        // Not being able to find a Node should not necessarily be an error
+        // Should this take PermanentId? Can the node Id shift out from underneath me?
+
+        // Take the dirty node list (replacing it with an empty Vec)
+        // Reverse the list and then remove duplicates
+        // (can't use Vec::dedup since that only removes consecutive duplicates)
+        let mut dirty = std::mem::take(&mut self.dirty_nodes);
+        dirty.reverse();
+        let mut node_list: Vec<_> = dirty.iter().unique().collect();
+
+        // Pop elements from back and update their size and Cid
+        while let Some(node_id) = node_list.pop() {
+            // Because of the work above we can assume that once we get here all of a nodes children are up to date
+            // The only error we can get from [`Self::by_id_mut`] is [`OperationError::InternalCorruption()`]
+            // which indicates the `NodeId` was not found. In our case that is fine, it just means it was deleted and
+            // we don't have to do anything
+            let node = self.by_id(*node_id);
+            if node.is_err() {
+                continue;
+            }
+            let node = node.unwrap();
```

Comment:
I'm really not fan of unwraps in production code, expect is always better forcing you to include a message if it panics but even that should be used extremely rarely. If there is some kind of internal corruption issue I'd much rather the encoding fail with a useful error message that is actionable rather than unwind the whole program.

Reply:
Agreed, I'm happy to make this an `expect`. I only used it along with the `is_err` check above to avoid deeply nesting the logic.  I originally used an `if let Some(node) = self.by_id(8node_id)` but that forces all of the code to update size (and the yet to be written code to update CID) to be nested into that `if` block. All that said if we remove nodes from the dirty list if they are removed from the filesystem as you suggest that should all be moot, and we can just use the `?` operator

Reply:
Yeah this should return an error if a node isn't found in this loop (`?` or if necessary generating an InternalCorruption error directly).